File: reinforcementLearningModels/Python/RLModel_RandomWalk_Python/RL_RandomWalk.py
#Random Walk Reinforcement Learning Model

#Clean environment
#from IPython import get_ipython
#get_ipython().magic('reset -f') #Clears variables
#get_ipython().magic('clear') #Clears console

#Import 
import numpy as np
import pandas as pd
import RL_RandomWalk_Functions as rl

#Setup
#Determine Target and Reward
pos_locs = np.array([np.arange(20),np.arange(20)])
np.random.shuffle(pos_locs[0,])
np.random.shuffle(pos_locs[1,])

start = np.array([pos_locs[0,0],pos_locs[1,0]])
end = np.array([pos_locs[0,-1],pos_locs[1,-1]])

board_template = np.zeros((20,20))
board_template[start[0],start[1]] = 1
board_template[end[0],end[1]] = -1

board_values = np.zeros((20,20))
board_values[end[0],end[1]] = 1

#Parameters
parameters = pd.DataFrame([np.zeros(9)],columns = ['target_x','target_y','temperature','learning_rate','discount','epsilon','responseSolutions','learningSolutions','steps'])
parameters['target_x'] = end[0]
parameters['target_y'] = end[1]
parameters['temperature'] = .1
parameters['learning_rate'] = .2
parameters['discount'] = .99
parameters['epsilon'] = .1
parameters['responseSolutions'] = 'SoftMax' #Can be 'Greedy' or 'SoftMax'
parameters['learningSolutions'] = 'SARSA' #Can be 'Q' or 'SARSA'
parameters['steps'] = np.nan

for walk in range (0,100):
    
    if walk % 100 == 0:
        print('Walk: '+ str(walk))
    
    board_state = np.zeros((20,20))
    board_state[start[0],start[1]] = 1

    x = start[0]
    y = start[1]
    
    ongoing = 1
    step = 0
    while ongoing:
        
        ## Select Action
        #Determine action of current state
        action,update_value = rl.select_action(x,y,board_values,parameters)
        
        #Determine Next State
        x2,y2 = rl.next_state(x,y,action)
        
        ## Update Model
        #Update Values
        board_values = rl.update_values(x,y,x2,y2,board_values,update_value,parameters)
        
        #Update Board State
        x,y,board_state,ongoing = rl.update_state(x,y,x2,y2,board_state,parameters)
        
        #Update step for model assessment
        step += 1
                
    ## Plot data
    parameters['steps'] = [np.append(parameters['steps'][0],step)]    
rl.plot_walk(board_template,board_state,board_values,parameters)

File: reinforcementLearningModels/Matlab/RLModel_RandomWalk/RL_RandomWalk.m
clear all; clc;

%% Initiation

%Determine board and starting/ target locations
board_values = ones(20,20)*0;
rand_spots = [Shuffle(1:20);Shuffle(1:20)];
board_values(rand_spots(1,1),rand_spots(2,1)) = 1;

%Create platform for model location
board_template = zeros(20,20);
board_template(rand_spots(1,end),rand_spots(2,end)) = -1;
board_template(rand_spots(1,1),rand_spots(2,1)) = 1;

%Model Parameters
parameters.target = [rand_spots(1,1),rand_spots(2,1)];
parameters.learning_rate = .7; %Learning Rate
parameters.temperature = .5; %Exploration rate of SoftMax solution
parameters.discount = .99; %Degree to which future rewards have influence
parameters.epsilon = .1; %Exploration rate of greedy solution
parameters.responseSolutions = 'SoftMax'; %Options are 'Greedy' and 'SoftMax'
parameters.learningSolutions = 'SARSA'; %Q or SARSA
parameters.Steps = []; %Create empty variable to track number of steps to target

%% Run
for walk = 1:100
    %Reset board
    ongoing = 1;
    board_state = zeros(20,20);
    board_state(rand_spots(1,end),rand_spots(2,end)) = 1;
    [x,y] = find(board_state==1);
    step = 0;
    while ongoing
        %% Select action
        %Determine action of current state
        [action, update_value] = select_action(x,y,board_values,parameters);
        
        %Determine next state
        [x2,y2] = next_state(x,y,action);
        
        %% Update Model
        %Update Values
        [board_values] = update_values(x,y,x2,y2,board_values,update_value,parameters);
        
        %Update Board State
        [ongoing,x,y,board_state] = update_state(x,y,x2,y2,board_state,parameters);
        
        %Update Step for Model Assessment
        step = step + 1;
    end
    
    %% Plot Output
    parameters.Steps = [parameters.Steps,step];
    plot_walk(board_template,board_values,parameters)
end

